src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -89,22 +114,30 @@ public SoftwareSpi(int clk, int miso, int mosi, int cs, SpiConnectionSettings se
             else
             {
                 _bitTransfer = new ScopeData(
+                    enter: () => { },
                     exit: () =>
                     {
                         _controller.Write(_clk, !idle);
                         _controller.Write(_clk, idle);
                     });
             }
 
-            _chipSelect = new ScopeData(
-                enter: () =>
-                {
-                    _controller.Write(_cs, _settings.ChipSelectLineActiveState);
-                },
-                exit: () =>
-                {
-                    _controller.Write(_cs, !(bool)_settings.ChipSelectLineActiveState);
-                });
+            if (_cs != -1)
+            {
+                _chipSelect = new ScopeData(
+                    enter: () =>
+                    {
+                        _controller.Write(_cs, _settings.ChipSelectLineActiveState);
+                    },
+                    exit: () =>
+                    {
+                        _controller.Write(_cs, !(bool)_settings.ChipSelectLineActiveState);
+                    });
+            }
+            else
+            {
+                _chipSelect = new ScopeData(() => { }, () => { });
```

Comment:
might be worth to actually make constructor default to null and replace null with nop internally, that won't pollute the code and still get rid of null checks

Reply:
Yea I agree, changed.